### PR TITLE
[BUGFIX] Resolve column types against the generic sqlalchemy namespace when the dialect module lacks them

### LINKED ISSUE #12215: [BUG] ExpectColumnValuesToBeOfType rejects the type name it reports: a type the dialect module does not export silently becomes a guaranteed failure
## What's wrong

`ExpectColumnValuesToBeOfType` and `ExpectColumnValuesToBeInTypeList` resolve the user's `type_` string by attribute lookup on the **SQLAlchemy dialect module** (`great_expectations/expectations/type_comparison.py:208`, `getattr(type_module, expected_type)`). When the dialect module does not export that name, the `AttributeError` is swallowed at DEBUG level (`:211`), the candidate-type list comes back empty, and the comparison at `:116` / `:156` becomes `isinstance(actual_column_type, ())` — which is unconditionally `False`. An unresolvable type name therefore produces a guaranteed failure that is indistinguishable from a genuine type mismatch: no exception, no warning, `raised_exception: false`.

On Oracle this makes the Expectation reject the very type name it reports. A column created as `INTEGER` is reflected as a generic `sqlalchemy.INTEGER`, so `observed_value` is `"INTEGER"` — but `sqlalchemy.dialects.oracle` exports no `INTEGER`, so asking `type_="INTEGER"` fails. There is no name a user can supply that passes, because the one GX itself reports is unresolvable.

## Impact

- **Fails closed, and lies about why.** A correct assertion on a correctly typed column reports failure, with an `observed_value` equal to the type asked for. The only clue is a DEBUG log line.
- On Oracle the two Expectations are effectively unusable for any type the dialect module does not re-export (`INTEGER` is one; the module exports its own `NUMBER`, `VARCHAR2`, etc.).
- On every dialect taking the `isinstance` path — Athena, BigQuery, ClickHouse, Dremio, Hive, MySQL, Oracle, Redshift, SingleStore, SQLite, Teradata, Vertica — a typo in `type_` (`"INTGER"`) reports as an ordinary mismatch rather than an error. Confirmed on SQLite: `type_="INTGER"` → `success=False, observed_value='INTEGER', raised_exception=False`.
- Dialects in `CASE_INSENSITIVE_DIALECTS` (`:53` — Databricks, PostgreSQL, Snowflake, SQL Server, Trino) take a string-comparison branch and are unaffected.
- Present since the lookup was introduced (`0.8.0a1`, 2019) through `1.23.0`. Workaround: none on Oracle for a generic type name.

## Reproduction

Reproduced on `1.23.0` (`673082d4b`), Python 3.11, against the repository's Oracle container (`assets/docker/oracle`).

Add to `tests/integration/data_sources_and_expectations/expectations/test_expect_column_values_to_be_of_type.py` (import `OracleDatasourceTestConfig` alongside the module's other configs):

```python
@parameterize_batch_for_data_sources(
    data_source_configs=[OracleDatasourceTestConfig()],
    data=DATA,
)
def test_success_for_type__INTEGER_oracle(batch_for_datasource: Batch) -> None:
    """Oracle reflects this column as a generic ``sqlalchemy.INTEGER`` and reports it as such.

    ``sqlalchemy.dialects.oracle`` does not export ``INTEGER``, so the expected-type lookup on
    the dialect module comes back empty and ``isinstance(value, ())`` is unconditionally False:
    the expectation rejects the very type name it reports as ``observed_value``.
    """
    result = batch_for_datasource.validate(
        gxe.ExpectColumnValuesToBeOfType(column=INTEGER_COLUMN, type_="INTEGER")
    )
    assert result.result["observed_value"] == "INTEGER"
    assert result.success, result.result
```

```bash
pytest tests/integration/data_sources_and_expectations/expectations/test_expect_column_values_to_be_of_type.py::test_success_for_type__INTEGER_oracle -m oracle
```

**Observed** (first assertion passes — the column *is* reported as `INTEGER` — the second fails):

```text
E   AssertionError: {'observed_value': 'INTEGER'}
E   assert False
  "result": {"observed_value": "INTEGER"},
  "exception_info": {"raised_exception": false, "exception_traceback": null, "exception_message": null}
```

**Expected:**

```text
success=True   # the column is INTEGER, and INTEGER was asked for
```

## Requirements

1. When `type_` (or a `type_list` member) names a type the dialect module does not export but the generic `sqlalchemy` namespace does, the comparison must resolve it against the generic namespace — so `ExpectColumnValuesToBeOfType(type_="INTEGER")` on an Oracle `INTEGER` column returns `success=True`, on any dialect that reflects the column as a generic SQLAlchemy type.
2. When a type name cannot be resolved in either namespace, the result must carry a clear exception naming the unresolvable type and the dialect (`raised_exception: true`), on every `isinstance`-path dialect — never a `success=False` that looks like a data mismatch. A typo must be told apart from a type mismatch.
3. Results for type names the dialect module *does* export must be unchanged on every dialect, and the `CASE_INSENSITIVE_DIALECTS` string-comparison branch must be unchanged.
4. The public signatures and parameters of both Expectations must not change.
5. The two behaviours above must each have a regression case in `test_expect_column_values_to_be_of_type.py` and `test_expect_column_values_to_be_in_type_list.py`: Oracle with `type_="INTEGER"` (passes), and an unresolvable name on at least one `isinstance`-path dialect (raises).

**Out of scope:** the LIKE-pattern allow-list on Oracle and SingleStore (#12214 — the same "dialect module is not a complete namespace" mistake, different code path); Snowflake regex anchoring (#12213); dialect-specific type vocabularies (ClickHouse's `Int32`/`String` naming is a separate question); the new schema-level `ExpectColumnTypeToBe` (#12186), which is converging on the same "raise clearly on an unknown type" rule independently.

## Context

- **Root cause:** `type_comparison.py:176-213` (`_get_potential_sqlalchemy_types`) — `getattr(type_module, expected_type)` at `:208`, swallowed at `:211`, then `isinstance(actual_column_type, tuple(types))` at `:116` (single type) and `:156` (type list) with an empty tuple. `observed_value` and the comparison use different sources: `type(actual_column_type).__name__` versus a lookup on the dialect module, so they can disagree exactly as shown above.
- **Fix shapes, each verified to turn the test above green on Oracle at `673082d4b`:**
  - **A** (requirement 1): after the dialect-module lookup, if `types` is still empty and `hasattr(sa, expected_type)`, append `getattr(sa, expected_type)`. Two lines at `:212-213`, where the empty list is already detected and only logged.
  - **C** (alternative, implementer's call): add `GXSqlDialect.ORACLE` to `CASE_INSENSITIVE_DIALECTS` at `:53`, so Oracle takes the string-comparison branch the five listed dialects already use. One line; also makes `NUMBER`, `VARCHAR2`, etc. compare by name. It changes Oracle's comparison from `isinstance` to case-insensitive string equality, which is a wider semantic shift than requirement 1 needs — and it does nothing for requirement 2 on the other eleven `isinstance`-path dialects. Reasonable if the implementer argues for it; A + B is the narrower shape.
- **B** (requirement 2) is independent of A/C: the empty-tuple `isinstance` at `:116` / `:156` should never be reached. Whether the exception is raised in `_get_potential_sqlalchemy_types` or checked by the two callers is the implementer's choice; either way it must surface in `exception_info`, not as a `success=False`.
- Ruled out: an Oracle reflection problem — `hasattr(sqlalchemy.dialects.oracle, "INTEGER")` is `False` and `isinstance(reflected_type, sqlalchemy.INTEGER)` is `True`, verified against the live container; the column really is a generic `INTEGER` and only the lookup namespace is wrong. Not verified: which other (dialect, generic type name) pairs are affected — Oracle/`INTEGER` is one instance, and dialect modules re-export whatever subset of generic types they choose.
- Also recorded in `tests/integration/data_sources_and_expectations/data_source_backlog.md` ("Fifteen candidates were measured…"): these two type cases are two of Oracle's six failures against the gallery-wide suite; the other four are #12214.

### PR DESCRIPTION
**Pull request by MannXo**
Resolves #12215

Changes:

**Fall back to the generic sqlalchemy namespace (requirement 1).** A dialect module re-exports only the subset of generic types it chooses to. `sqlalchemy.dialects.oracle` exports `NUMBER` and `VARCHAR2` but not `INTEGER`, while an Oracle `INTEGER` column is reflected as a generic `sqlalchemy.INTEGER`. When the dialect-module lookup finds nothing, `_get_potential_sqlalchemy_types` now tries the generic namespace before giving up.

**Raise on a name that resolves nowhere (requirement 2).** An empty candidate list made the comparison `isinstance(value, ())`, which is unconditionally `False`, so a typo was reported as an ordinary mismatch. An unresolvable name now raises `InvalidExpectationConfigurationError` naming the type and the dialect, which surfaces as `raised_exception: true` rather than a bare `success: false`.

I took shape A plus B rather than C. Adding Oracle to `CASE_INSENSITIVE_DIALECTS` would change Oracle's comparison from `isinstance` to string equality, which is wider than requirement 1 needs, and it would do nothing for requirement 2 on the other eleven `isinstance`-path dialects.

**One place I deviated from the requirements, deliberately, and I would like your view on it.**

Requirement 2 reads as though any unresolvable name should raise. Applied to `ExpectColumnValuesToBeInTypeList` that breaks the library's own idiom, and the existing suite is what caught it: `test_success_complete` in `test_expect_column_values_to_be_in_type_list.py` validates against every backend with a single cross-backend list.

```python
type_list = ["INTEGER", "Integer", "int", "int64", "int32", "IntegerType", "_CUSTOM_DECIMAL"]
```

`int64` and `int32` are pandas names, `IntegerType` is Spark. On sqlite most of that list is unresolvable by design, and raising on the first one fails the test. So a name a given dialect cannot resolve is normal in a type list rather than a typo.

A type list therefore tolerates individual unresolvable names and raises only when nothing in the list resolves, which is exactly the case that produced the empty-tuple bug. The scalar `type_` keeps the strict behaviour requirement 2 asks for, since there is no cross-backend reading of a single name. If you would rather have strictness on both, say so and I will change it, though I think it would have to come with a way to express a cross-backend list.

**Two things I left alone**, both flagged rather than silently skipped. The BigQuery `GEOGRAPHY` branch deliberately yields no candidate types and emits an actionable `logger.warning` about installing the extra; I returned early there rather than convert that into an exception. And the Athena, ClickHouse and Redshift branches can yield `[None]` rather than an empty list, which is a separate latent problem that this change neither fixes nor worsens.

**On requirement 5, and this is the gap in this PR.** The unresolvable-name half has regression cases in both files, on sqlite, and I ran them. The Oracle half is not covered: I have no Oracle container available, so rather than commit a test I could not execute I left it out. `TestPartialDialectModuleFallback` in `tests/expectations/test_type_comparison.py` covers the same code path with a stub dialect module that exports `NUMBER` but not `INTEGER`, which reproduces the Oracle condition without the database, but it is not the integration test you asked for. Happy to add the Oracle case from your reproduction if you would prefer it in even though it will first run in your CI.

Verification, all local:

- `invoke lint --no-fmt`, `invoke fmt --check`, `invoke type-check --ci --pretty`: all clean
- `pytest tests/expectations/test_type_comparison.py`: 857 passed
- the two integration files under `-m sqlite`: 4 passed
- full unit suite: 5228 passed. One unrelated failure in `test_azure_blob_storage_data_connector.py` reproduces identically on `develop` with my changes stashed

No RFC needed: bug fix, no public signature or parameter change.

#### AI disclosure

I used Claude Code assistively here. It helped me trace the lookup and the `isinstance` path through the source and drafted wording, which I reviewed and edited. The cross-backend list finding above came out of running the existing suite rather than from reading, and every claim in this description is checked against the code or a command I ran. I take responsibility for the change.

### CODE REVIEW

File: great_expectations/expectations/type_comparison.py [hunk 1]
```diff
@@ -209,8 +234,18 @@ def _get_potential_sqlalchemy_types(
...
     except AttributeError:
         logger.debug(f"Unrecognized type: {expected_type}")
-    if len(types) == 0:
-        logger.debug("No recognized sqlalchemy types in type_list for current dialect.")
+
+    if not types:
+        # A dialect module re-exports only the subset of generic types it chooses to.
+        # Oracle exports NUMBER and VARCHAR2 but not INTEGER, yet reflects an INTEGER
+        # column as a generic sqlalchemy.INTEGER.
+        generic_type = getattr(sa, expected_type, None)
```
Comment (joshua-stauffer) [inline great_expectations/expectations/type_comparison.py:242]: The generic fallback accepts any attribute of `sqlalchemy`'s top-level namespace, not just type classes. A name that collides with a non-class export (e.g. `text`, `cast`, `func`, `select`, `table`) lands in `types`, and `isinstance(actual_column_type, tuple(types))` (lines 117, 159) raises `TypeError: isinstance() arg 2 must be a type...`, naming neither the type nor the dialect. On sqlite, `type_="text"` previously resolved to nothing and returned a clean `success: false`; now it crashes, bypassing the `InvalidExpectationConfigurationError` path this PR adds. `type_list` is worse: one bad name poisons the whole candidate tuple, so `type_list=["text", "INTEGER"]` crashes on any dialect missing that name even though `INTEGER` resolves fine. `expect_column_values_to_be_of_type.py:376` already guards this exact `getattr`-into-a-large-namespace pattern on the pandas side (`getattr(pd, expected_type)` only `if isinstance(pd_type, type)`); the SQL fallback has no equivalent check. Restrict it to `isinstance(generic_type, type) and issubclass(generic_type, sa.types.TypeEngine)` and treat anything else as unresolved.

_Review powered by GX Maintainer Bot 🦖_

File: great_expectations/expectations/type_comparison.py [hunk 2]
```diff
@@ -150,9 +151,11 @@ def compare_column_type_list(
         for type_ in expected_types_list:
             types.extend(
                 _get_potential_sqlalchemy_types(
-                    execution_engine=execution_engine, expected_type=type_
+                    execution_engine=execution_engine, expected_type=type_, strict=False
                 )
             )
+        if not types:
+            raise _unresolvable_type_error(execution_engine, expected_types_list)
```
Comment (joshua-stauffer) [inline great_expectations/expectations/type_comparison.py:158]: This new raise defeats the BigQuery GEOGRAPHY carve-out the same diff deliberately adds, in the list-expectation path. `_get_potential_sqlalchemy_types` returns an empty list early at line 217 specifically so a missing `sqlalchemy-bigquery[geography]` extra warns rather than raises, but `type_list=["GEOGRAPHY"]` on BigQuery leaves the aggregate `types` empty here, so it now raises `InvalidExpectationConfigurationError` telling the user the name "could not be resolved against the bigquery dialect module or the generic sqlalchemy namespace. Check the spelling, or use a type name the dialect supports" — when the name is spelled correctly, is supported, and the real cause is the uninstalled optional extra the immediately preceding `logger.warning` correctly diagnoses. It is also asymmetric: the scalar `type_="GEOGRAPHY"` path on the same engine still returns `success: false` (the early return precedes the strict raise there). BigQuery users without the geography extra get a behavior change (warning + failure → raised exception with a misleading cause) that the PR body states was deliberately avoided.

_Review powered by GX Maintainer Bot 🦖_

File: great_expectations/expectations/type_comparison.py [hunk 3]
```diff
@@ -209,8 +234,18 @@ def _get_potential_sqlalchemy_types(
...
-    if len(types) == 0:
-        logger.debug("No recognized sqlalchemy types in type_list for current dialect.")
+
+    if not types:
+        # A dialect module re-exports only the subset of generic types it chooses to.
+        # Oracle exports NUMBER and VARCHAR2 but not INTEGER, yet reflects an INTEGER
+        # column as a generic sqlalchemy.INTEGER.
+        generic_type = getattr(sa, expected_type, None)
+        if generic_type is not None:
+            types.append(generic_type)
```
Comment (joshua-stauffer) [inline great_expectations/expectations/type_comparison.py]: This fallback silently widens type matching from exact dialect-type names to generic base classes on every isinstance-path dialect, not just Oracle — a behavioral change to existing users' validation results, in the permissive direction, on upgrade, not covered by the PR body's "no public signature or parameter change" claim.

Concretely on MySQL: `sqlalchemy.dialects.mysql` exports `INTEGER`/`BIGINT` but no CamelCase `Integer`, so `ExpectColumnValuesToBeOfType(type_="Integer")` against a `BIGINT` column used to hit `AttributeError` → empty list → `isinstance(x, ())` → `success: false`. With this line it resolves to `sa.Integer`, and `mysql.BIGINT` derives from `sqltypes.BIGINT` → `BigInteger` → `Integer`, so the same expectation now reports `success: true`.

It also creates a cross-dialect disagreement: on the five `CASE_INSENSITIVE_DIALECTS`, `type_="Integer"` stays a literal name comparison and still fails, while on mysql/sqlite/oracle it now succeeds. No test in this diff covers the abstract-base-class case.

_Review powered by GX Maintainer Bot 🦖_

File: great_expectations/expectations/type_comparison.py [hunk 4]
```diff
@@ -209,8 +234,18 @@ def _get_potential_sqlalchemy_types(
...
+        elif strict:
+            raise _unresolvable_type_error(execution_engine, [expected_type])
```
Comment (joshua-stauffer) [inline great_expectations/expectations/type_comparison.py]: Raising here makes validation outcomes depend on the installed SQLAlchemy major version, which this repo still supports on both sides (`reqs/requirements-dev-sqlalchemy1.txt` pins `sqlalchemy<2.0.0`; `pyproject.toml` registers a `sqlalchemy_version_compatibility` marker). The fallback resolves names from the installed `sqlalchemy` top-level namespace, and that set grew in 2.0 — e.g. `Double` is exported at the top level only from 2.0 onward. On an isinstance-path dialect whose module doesn't re-export it, `type_="Double"` resolves under SQLAlchemy 2.x but raises under 1.x, telling the user to "Check the spelling" of a correctly spelled, dialect-supported name. Before this change both versions returned `success: false`, so the version difference was inert; this turns it into an exception the user can't act on from the message.

_Review powered by GX Maintainer Bot 🦖_

File: great_expectations/expectations/type_comparison.py [hunk 5]
```diff
@@ -190,6 +212,9 @@ def _get_potential_sqlalchemy_types(
                 + "To install support, please run:"
                 + "  $ pip install 'sqlalchemy-bigquery[geography]'"
             )
+            # Missing optional support, not an unresolvable name: the warning above is
+            # actionable, so return empty rather than raising below.
+            return types
```
Comment (joshua-stauffer) [inline great_expectations/expectations/type_comparison.py:217]: The BigQuery GEOGRAPHY-without-extra branch returns `types = []` before the strict/generic-fallback block runs, bypassing the strict-raise behavior this PR makes the default (`strict: bool = True`) for `_get_potential_sqlalchemy_types`. Calling `compare_column_type` (the scalar path used by `ExpectColumnValuesToBeOfType`) with `expected_type="GEOGRAPHY"` on a BigQuery engine lacking the geography extra still returns `success=False` instead of raising — the pre-existing defect requirement 2 is meant to eliminate, confined to this dialect/type combination. The PR body calls this a deliberate, known deviation, but nothing in the diff tests it: `TestPartialDialectModuleFallback` covers only the Oracle generic-fallback path via a stub module, and the sqlite/mysql tests never touch the BigQuery branch. It's testable without a live BigQuery connection — the branch (lines 205-209) is gated on `execution_engine.engine.dialect.name` and the import-time constant `BIGQUERY_GEO_SUPPORT` (`great_expectations/compatibility/bigquery.py`), so a test needs an engine stub returning `"bigquery"` for `.engine.dialect.name` plus a patched/false `BIGQUERY_GEO_SUPPORT` — but no such test exists, so a later refactor could silently flip this behavior either way with nothing catching it.

_Review powered by GX Maintainer Bot 🦖_

File: great_expectations/expectations/type_comparison.py
Comment (joshua-stauffer) [inline great_expectations/expectations/type_comparison.py:1]: The new unresolvable-name strictness is dialect-partial, so the same misspelled config now behaves differently per backend, and nothing in the change accounts for it.

It cannot fire on ClickHouse. `dialect_module` there is `clickhouse_sqlalchemy.drivers.base`, so the `elif` at line 227 calls `get_clickhouse_sqlalchemy_potential_type`, and for an unrecognized spec `ClickHouseDialect._get_column_type` warns and returns the class `sqltypes.NullType`. `types` therefore ends up `[NullType]` — non-empty — so both `if not types` guards (lines 157 and 238) are skipped and the comparison degrades to an `isinstance` against `NullType`, i.e. a silent `success: false` again.

At the other extreme, Athena gets more strictness than the PR body describes rather than less: `SqlAlchemyExecutionEngine.__init__` sets `dialect_module` only for POSTGRESQL/MYSQL/SQLITE/ORACLE/SQL_SERVER/SNOWFLAKE/DREMIO/REDSHIFT/BIGQUERY/TERADATASQL/TRINO/CLICKHOUSE/DATABRICKS/SINGLESTOREDB, and `awsathena` falls to `dialect_module = None`. `_get_dialect_type_module` then returns bare `sa` (the `None` check at line 270, returning at line 272), the `pyathena` branch at line 218 is never entered, and every name is resolved against the generic sqlalchemy namespace — so a legitimate Athena-specific type name that `sa` does not export now raises `InvalidExpectationConfigurationError` where it previously returned a clean `success: false`.

Can we bring Clickhouse and Athena behavior in line and ensure both cases are under test?

_Review powered by GX Maintainer Bot 🦖_

File: tests/integration/data_sources_and_expectations/expectations/test_expect_column_values_to_be_of_type.py
Comment (joshua-stauffer) [inline tests/integration/data_sources_and_expectations/expectations/test_expect_column_values_to_be_of_type.py:1]: Issue #12215 requires two regression cases in both `test_expect_column_values_to_be_of_type.py` and `test_expect_column_values_to_be_in_type_list.py`: an Oracle `type_="INTEGER"` passing case (per the issue's own `OracleDatasourceTestConfig` repro) and an unresolvable-name-raises case on at least one isinstance-path dialect. This diff adds only the unresolvable-name case (sqlite) to both files — neither integration file references Oracle. The generic-fallback case is covered only by a unit test against a stub dialect module, not an integration test against the real Oracle backend the issue called for.

_Review powered by GX Maintainer Bot 🦖_

File: great_expectations/expectations/type_comparison.py [hunk 6]
```diff
@@ -173,9 +188,16 @@ def _compare_type_string(actual_column_type: Any, expected_type: str) -> bool:
     return str(actual_column_type).casefold() == expected_type.casefold()
 
 
-def _get_potential_sqlalchemy_types(
-    execution_engine: SqlAlchemyExecutionEngine, expected_type: str
+def _get_potential_sqlalchemy_types(  # noqa: C901
```
Comment (joshua-stauffer) [inline great_expectations/expectations/type_comparison.py:191]: we should refactor rather than add a noqa here

File: tests/integration/data_sources_and_expectations/expectations/test_expect_column_values_to_be_of_type.py
Comment (joshua-stauffer) [reply tests/integration/data_sources_and_expectations/expectations/test_expect_column_values_to_be_of_type.py:1]: I did see your note in the PR body with regards to the oracle container - we do want a full integration test here. Fine to have the test run for the first time in CI, but you can also use `assets/docker/oracle/docker-compose.yml` locally to get direct feedback.

**Review by joshua-stauffer — CHANGES_REQUESTED** [review]
Thanks for this PR @MannXo! this is a tricky change, and your approach with the `strict` param is good. I've noted some edge cases we need to address prior to merging, along with some additional test coverage.
